### PR TITLE
Closes #7062: Add nested menu UI for web extensions actions

### LINKED ISSUE #7062: Add sub menu capability for nested menu
For https://github.com/mozilla-mobile/fenix/issues/8319

We need a menu item that can handle nested sub menu.

### PR DESCRIPTION
**Pull request by psymoon**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/menu/src/main/java/mozilla/components/browser/menu/WebExtensionBrowserMenuBuilder.kt [hunk 1]
```diff
@@ -21,24 +25,53 @@ class WebExtensionBrowserMenuBuilder(
     extras: Map<String, Any> = emptyMap(),
     endOfMenuAlwaysVisible: Boolean = false,
     private val store: BrowserStore,
+    private val webExtIconTintColorResource: Int = NO_ID,
+    private val onAddonsManagerTapped: () -> Unit = {},
```
Comment (Amejia481) [inline components/browser/menu/src/main/java/mozilla/components/browser/menu/WebExtensionBrowserMenuBuilder.kt:29]: nit: let's add some doc for the new properties

**Comment by Amejia481** [comment]
@psymoon  it's looking really good! 🦄 
I was wondering if we should have a message for when there are no add-ons, this doesn't block the pr in any case could be addressed in follow up.

image


Similar to what we do with https://github.com/mozilla-mobile/fenix/blob/e21352144b4f8f94453c55bf8ada4356eeddfde0/app/src/main/res/values/strings.xml#L73

File: components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserNestedMenuItemTest.kt [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        val textView = view.findViewById<TextView>(R.id.text)
+        val imageView = view.findViewById<AppCompatImageView>(R.id.image)
+
+        assertEquals("label", textView.text)
+        assertNotNull(imageView.drawable)
+        assertNotNull(imageView.imageTintList)
+    }
+
+    @Test
+    fun `menu item ImageText with with no iconTintColorResource must not have an imageTintList`() {
```
Comment (Amejia481) [inline components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserNestedMenuItemTest.kt:55]: nit: with with

**Comment by psymoon** [comment]
> @psymoon it's looking really good! 🦄
> I was wondering if we should have a message for when there are no add-ons, this doesn't block the pr in any case could be addressed in follow up if needed.
> 
> image
> 
> Similar to what we do with https://github.com/mozilla-mobile/fenix/blob/e21352144b4f8f94453c55bf8ada4356eeddfde0/app/src/main/res/values/strings.xml#L73

@Amejia481: @brampitoyo's request for this per slack:
"When there’s no add-on installed, can we not show this sub-menu and instead make the “Add-ons” menu item (one level up) deep-link to the Add-ons Manager? That’s the behaviour I’d prefer."

This is updated and will need re-review (specifically for `WebExtensionBrowserMenuBuilder`)

File: components/browser/menu/src/main/java/mozilla/components/browser/menu/item/BrowserNestedMenuItem.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import android.view.View
+import androidx.annotation.VisibleForTesting
+import mozilla.components.browser.menu.BrowserMenu
+import mozilla.components.browser.menu.BrowserMenuItem
+
+/**
+ * An abstract menu item for handling nested sub menu items on view click.
+ *
+ * @param subMenu Target sub menu to be shown when this menu item is clicked.
```
Comment (csadilek) [inline components/browser/menu/src/main/java/mozilla/components/browser/menu/item/BrowserNestedMenuItem.kt:15]: nit: `@property`

File: components/browser/menu/src/main/java/mozilla/components/browser/menu/item/WebExtensionBrowserNestedMenuItem.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * A menu item for displaying text and an image icon with nested sub menu capabilities.
+ *
+ * @param label The visible label of this menu item.
+ * @param imageResource ID of a drawable resource to be shown as icon.
+ * @param iconTintColorResource Optional ID of color resource to tint the icon.
+ * @param textColorResource Optional ID of color resource to tint the text.
+ */
+@Suppress("LongParameterList")
+class WebExtensionBrowserNestedMenuItem(
```
Comment (csadilek) [inline components/browser/menu/src/main/java/mozilla/components/browser/menu/item/WebExtensionBrowserNestedMenuItem.kt:30]: nit: Unlike `WebExtensionsBrowserMenuItem`, there's actually nothing specific to web extensions in this nested menu item implementation? It doesn't reference anything related to extensions. Should it just be called `BrowserNestedMenuItem` or `DefaultBrowserNestedMenuItem` if you want to keep the name of the abstract class?

File: components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuBuilderTest.kt [hunk 5]
```diff
@@ -4,28 +4,54 @@
...
 import org.junit.Test
 import org.junit.runner.RunWith
-import org.mockito.Mockito
 
 @RunWith(AndroidJUnit4::class)
 class WebExtensionBrowserMenuBuilderTest {
 
     @Test
-    fun `web extension is added at the start when appendExtensionActionAtStart is true`() {
+    fun `WHEN there are no web extensions installed THEN add-ons menu item invokes onAddonsManagerTapped`() {
```
Comment (csadilek) [inline components/browser/menu/src/test/java/mozilla/components/browser/menu/WebExtensionBrowserMenuBuilderTest.kt:32]: nit: when there are no web extension actions...

(We could have add-ons installed that don't have browser actions :)

File: components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserNestedMenuItemTest.kt [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class WebExtensionBrowserNestedMenuItemTest {
```
Comment (csadilek) [inline components/browser/menu/src/test/java/mozilla/components/browser/menu/item/WebExtensionBrowserNestedMenuItemTest.kt:28]: nit: newline.

File: docs/changelog.md [hunk 7]
```diff
@@ -38,6 +38,11 @@ permalink: /changelog/
 * **service-glean**
   * BUGFIX: Fix a race condition that leads to a `ConcurrentModificationException`. [Bug 1635865](https://bugzilla.mozilla.org/1635865)
 
+* **browser-menu**
+  * ⚠️ **This is a breaking change**: `WebExtensionBrowserMenuBuilder` now returns as a sub menu entry for add-ons. The sub
```
Comment (csadilek) [inline docs/changelog.md:42]: 👍 let's get a PR up for Fenix too :)

**Comment by psymoon** [comment]
bors r=Amejia481, csadilek

**Pull request merged**